onnx/defs/math/defs.cc

```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
           defs::math::utils::MatMulShapeInference(ctx, 0, 1);
         }));
 
+static const char* MatMulNBits_ver23_doc = R"DOC(
+MatMulNBits is a MatMul with weight quantized with N bits (e.g., 2, 3, 4, 5, 6, 7). It does Matrix Multiplication like
+[MatMul](#matmul) with differences:
+
+1. Input `B` is a 2D constant Matrix. Its input feature count and output feature count are specified by attributes 'K'
+    and 'N'.
+2. Input `B` is quantized with x bits which is specified by attribute 'bits'. It is quantized blockwisely along
+    dimension 0 (e.g. column) with block size specified by attribute block_size. And block_size is not an arbitrary
+    number and must be a power of 2 and not smaller than 16, like 16, 32, 64, 128,..
+3. Input `B`'s scale and zero point are specified by input scales and zero_points.
+
+    Input `B` is stored as uint8_t with shape: `[N][n_blocks_per_col][blob_size]`
+
+    in which:
+      - `n_blocks_per_col` = `(K + block_size - 1) / block_size`
+      - `blob_size` = `CeilDiv(block_size * bits, 8)`
+
+    For all bits from 2-8, a row of data is tightly packed and represented by uint8_t.
+    The bit packing specified for [4 bit integer types](https://onnx.ai/onnx/technical/int4.html) is followed.
+    The sub-byte values always start from the LSB of the byte.
+
+    - for 2,4,8 bits, 4x2bit,2x4bit,1x8bit are stored in one `uint8_t`.
+      ```
+        4bit example: with 3x4bit values
+        Value A: 4 bits (Bits 0:3 in byte 0)
+        Value B: 4 bits (Bits 4:7 in byte 0)
+        Value C: 4 bits (Bits 0:3 in byte 1)
+        Packed Bit Layout:
+        Byte 0: [ B B B B A A A A ] -> A[0:3], B[4:7]
+        Byte 1: [ . . . .  C C C C] -> C[0:3]
+      ```
+
+    - for 3,5,6,7 bits, 32x3bit,32x5bit,16x6bit,32x7bit are stored in 12x`uint8_t`, 20x`uint8_t`, 12x`uint8_t`,
```

Comment:
For clarification: so, this packing is different from what onnxruntime does? (May require repacking if we migrate models using the onnxruntime op to this op.)

Reply:
The 4-Bit integer packing is identical to the packing specified for int4 by ONNX specification and ONNX Runtime.

The packing done for all the other N-Bit types is specified this way to match the specification for int4. Although onnxruntime specifies the com.microsoft.MatMulNBits they currently only have an implementation for the 4-bit variant at this time. I have tested the packing as specified for 4-bit data in ONNX Runtime and get the correct results.

My first draft had the order different `[AAAABBBB]` for Byte1 and `[CCCC....]` for Byte2 found out though testing it was not packed as expected. That is when I found the [4-bit integer types](https://onnx.ai/onnx/technical/int4.html) link and matched the description to that.